smt.go

```diff
@@ -138,9 +138,10 @@ func (smt *SparseMerkleTree) HasForRoot(key, root []byte) (bool, error) {
 // Update sets a new value for a key in the tree, and sets and returns the new root of the tree.
 func (smt *SparseMerkleTree) Update(key []byte, value []byte) ([]byte, error) {
 	newRoot, err := smt.UpdateForRoot(key, value, smt.Root())
-	if err == nil {
-		smt.SetRoot(newRoot)
+	if err != nil {
+		return newRoot, err
```

Comment:
```suggestion
		return nil, err
```
`newRoot` is `nil` anyways, but it's not clear at first glance (you need to check few functions).